www/js/survey/input-matcher.ts

```diff
@@ -0,0 +1,243 @@
+    start_ts: number,
+    end_ts: number,
+    enter_ts: number,
+    exit_ts: number,
+    duration: number,
+    getNextEntry: any
+}
+
+
+const EPOCH_MAXIMUM = 2**31 - 1;
+
+export const fmtTs = (ts_in_secs: number, tz: string): string | null => DateTime.fromSeconds(ts_in_secs, {zone : tz}).toISO();
+
+export const printUserInput = (ui: UserInputForTrip): string => `${fmtTs(ui.data.start_ts, ui.metadata.time_zone)} (${ui.data.start_ts}) -> 
+${fmtTs(ui.data.end_ts, ui.metadata.time_zone)} (${ui.data.end_ts}) ${ui.data.label} logged at ${ui.metadata.write_ts}`;
+
+export const validUserInputForDraftTrip = (trip: Trip, userInput: UserInputForTrip, logsEnabled: boolean): boolean => {
+    if(logsEnabled) {
+        logDebug(`Draft trip:
+            comparing user = ${fmtTs(userInput.data.start_ts, userInput.metadata.time_zone)}
+            -> ${fmtTs(userInput.data.end_ts, userInput.metadata.time_zone)}
+            trip = ${fmtTs(trip.start_ts, userInput.metadata.time_zone)}
+            -> ${fmtTs(trip.end_ts, userInput.metadata.time_zone)}
+            checks are (${userInput.data.start_ts >= trip.start_ts}
+            && ${userInput.data.start_ts < trip.end_ts}
+            || ${-(userInput.data.start_ts - trip.start_ts) <= 15 * 60})
+            && ${userInput.data.end_ts <= trip.end_ts}
+        `);
+    }
+
+    return (userInput.data.start_ts >= trip.start_ts && userInput.data.start_ts < trip.end_ts
+    || -(userInput.data.start_ts - trip.start_ts) <= 15 * 60) && userInput.data.end_ts <= trip.end_ts;
+}
+
+export const validUserInputForTimelineEntry = (tlEntry: TlEntry, userInput: UserInputForTrip, logsEnabled: boolean): boolean => {
+    if (!tlEntry.origin_key) return false;
+    if (tlEntry.origin_key.includes('UNPROCESSED')) return validUserInputForDraftTrip(tlEntry, userInput, logsEnabled);
+
+    /* Place-level inputs always have a key starting with 'manual/place', and
+        trip-level inputs never have a key starting with 'manual/place'
+       So if these don't match, we can immediately return false */
+    const entryIsPlace = tlEntry.origin_key === 'analysis/confirmed_place';
+    const isPlaceInput = (userInput.key || userInput.metadata.key).startsWith('manual/place');
+    
+    if (entryIsPlace !== isPlaceInput) return false;
+
+    let entryStart = tlEntry.start_ts || tlEntry.enter_ts;
+    let entryEnd = tlEntry.end_ts || tlEntry.exit_ts;
+    
+    if (!entryStart && entryEnd) {
+      /* if a place has no enter time, this is the first start_place of the first composite trip object
+      so we will set the start time to the start of the day of the end time for the purpose of comparison */
+      entryStart = DateTime.fromSeconds(entryEnd).startOf('day').toUnixInteger();
+    }
+
+    if (!entryEnd) {
+        /* if a place has no exit time, the user hasn't left there yet
+        so we will set the end time as high as possible for the purpose of comparison */
+        entryEnd = EPOCH_MAXIMUM;
+    }
```

Comment:
When `tlEntry.end_ts` and `tlEntry.exit_ts` are equal to `0`, `entryEnd` is `falsy` value so it goes into this `if statement` and set `entryEnd` as `EPOCH_MAXIMUM`, and we get the wrong result. Should I handle the case when `tlEntry.end_ts` and `tlEntry.exit_ts` are equal to `0`?

Reply:
It is really great that you're considering edge cases like this!
`end_ts` and `exit_ts` are measurements in "Unix time" aka "Epoch seconds" https://en.wikipedia.org/wiki/Unix_time.

In the modern era, this should be a 10-digit number (like 1697473049)

If either of these measurements were 0, it would definitely be an invalid input. So I think in this case, it is fair to treat a value of 0 the same as if it was undefined; all falsy values will be treated the same way. So it is safe the way it is.